File: Go/Day01/day01-report_repair.go
package main

import (
	"fmt"
	"io/ioutil"
	"os"
	"strconv"
	"strings"
)

func main() {
	fmt.Println("Day 01: Report Repair\n=====================")

	expenseNumbers := readExpenseReport("RawData.txt")
	fmt.Printf("Read numbers: %q\n", expenseNumbers)

	fmt.Println("\nPart 1: Two numbers who add up to 2020, their product\n-----------------------------------------------------")
	solvePart1(expenseNumbers)

	fmt.Println("\nPart 2: Three numbers who add up to 2020, their product\n-------------------------------------------------------")
	solvePart2(expenseNumbers)
}

func readExpenseReport(filename string) []int {
	buf, err := ioutil.ReadFile(filename)
	if err != nil {
		fmt.Fprint(os.Stderr, "Error occurred when trying to read data from file: ", err)
		os.Exit(1)
	}

	numberStrings := strings.Split(string(buf), "\r\n")
	//fmt.Printf("%q\n", numberStrings)
	//fmt.Println(strings.Join(numberStrings, ","))

	var res []int
	for _, n := range numberStrings {
		i, err := strconv.Atoi(n)
		if err != nil {
			fmt.Fprintf(os.Stderr, "Error parsing '%s' as integer: %q", n, err)
			continue
		}
		res = append(res, i)
	}

	return res
}

func solvePart1(numbers []int) {
	for i, n := range numbers {
		for _, m := range numbers[i+1:] {
			if m+n == 2020 {
				fmt.Printf("Found both numbers: %d + %d = %d\n", m, n, m+n)
				res := m * n
				fmt.Printf("Multiplication result: %d\n", res)
			}
		}
	}
}

func solvePart2(numbers []int) {
	for i, n := range numbers {
		for j, m := range numbers[i+1:] {
			for _, o := range numbers[i+j:] {
				if m+n+o == 2020 {
					fmt.Printf("Found a numbers: %d + %d + %d = %d\n", m, n, o, m+n+o)
					res := m * n * o
					fmt.Printf("Multiplication result: %d\n", res)
				}
			}
		}
	}
}

File: TypeScript/day01-report_repair.ts
import { readFileSync } from 'fs';
import { exit } from 'process';


function getElementsForSum(numbers: number[], count: number, targetSum: number): number[] {
    if (count == 2){
        // console.log(`Reached final case, only 2 number required to reach ${targetSum} from array of length ${numbers.length}`)
        for (const { i, v } of numbers.map((v, i) => ({ i, v }))){
            for (var j=i+1;+j < numbers.length; ++j){
                if (v + numbers[j] == targetSum){
                    return [ v, numbers[j] ]
                }
            }
        }
    } else {
        for (let _i in numbers){
            if (targetSum > numbers[_i] + count - 1){
                let res = getElementsForSum(numbers.slice(+_i+1, numbers.length), count-1, targetSum - numbers[+_i])
                if (res != null){
                    return [numbers[+_i]].concat(res)
                }
            }
        }
    }

    return null
}


console.log("Day 01: Report Repair\n=====================")

const lines = readFileSync('./day01-rawdata.txt', 'utf-8').split("\n")
console.log("Numbers from input:")
console.log(lines.map(function (s, _){ return +s }))

console.log("Part 1: Two numbers who add up to 2020, their product\n-----------------------------------------------------")
let resPart1 = getElementsForSum(lines.map(function (s, _){ return +s }), 2, 2020)
console.log(`Found both numbers: ${resPart1[0]} + ${resPart1[1]} = ${resPart1[0]+resPart1[1]}`)
console.log("Multiplication result: %d\n", resPart1.reduce(function (acc, val, ind){ return acc * val }, 1))

console.log("Part 2: Three numbers who add up to 2020, their product\n-------------------------------------------------------")
let resPart2 = getElementsForSum(lines.map(function (s, _){ return +s }), 3, 2020)
console.log(`Found all three numbers: ${resPart2[0]} + ${resPart2[1]} + ${resPart2[2]} = ${resPart2[0]+resPart2[1]+resPart2[2]}`)
console.log("Multiplication result: %d\n", resPart2.reduce(function (acc, val, ind){ return acc * val }, 1))






